conductor_api/src/keystore.rs

```diff
@@ -0,0 +1,763 @@
+use holochain_core_types::{
+    agent::Base32,
+    cas::content::Address,
+    error::{HcResult, HolochainError},
+    signature::Signature,
+};
+use holochain_dpki::{
+    key_blob::{BlobType, Blobbable, KeyBlob},
+    key_bundle::KeyBundle,
+    keypair::{generate_random_sign_keypair, EncryptingKeyPair, KeyPair, SigningKeyPair},
+    seed::Seed,
+    utils::{
+        decrypt_with_passphrase_buf, encrypt_with_passphrase_buf, generate_derived_seed_buf,
+        generate_random_buf, verify as signingkey_verify, SeedContext,
+    },
+    SEED_SIZE,
+};
+
+use holochain_sodium::{
+    pwhash::{ALG_ARGON2ID13, MEMLIMIT_INTERACTIVE, OPSLIMIT_INTERACTIVE},
+    secbuf::SecBuf,
+};
+
+use conductor::passphrase_manager::PassphraseManager;
+use holochain_dpki::{password_encryption::PwHashConfig, seed::SeedType};
+use std::{
+    collections::{BTreeMap, HashMap},
+    fs::File,
+    io::prelude::*,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+
+const PCHECK_HEADER_SIZE: usize = 8;
+const PCHECK_HEADER: [u8; 8] = *b"PHCCHECK";
+const PCHECK_RANDOM_SIZE: usize = 32;
+const PCHECK_SIZE: usize = PCHECK_RANDOM_SIZE + PCHECK_HEADER_SIZE;
+const KEYBUNDLE_SIGNKEY_SUFFIX: &str = ":sign_key";
+const KEYBUNDLE_ENCKEY_SUFFIX: &str = ":enc_key";
+
+pub enum Secret {
+    SigningKey(SigningKeyPair),
+    EncryptingKey(EncryptingKeyPair),
+    Seed(SecBuf),
+}
+
+enum KeyType {
+    Signing,
+    Encrypting,
+}
+
+#[allow(dead_code)]
+#[derive(Serialize, Deserialize)]
+pub struct Keystore {
+    passphrase_check: String,
+    secrets: BTreeMap<String, KeyBlob>,
+    #[serde(skip_serializing, skip_deserializing)]
+    cache: HashMap<String, Arc<Mutex<Secret>>>,
```

Comment:
curious `secrets` and `cache` are both indexed by the same id strings, why is one a btree and the other a hash map?

Reply:
`cache` is only held in memory -> sorting is irrelevant. `secrets` is persisted to files. `BTreeMap` makes sure that the sorting and hence the exact serialized representation is deterministic.

@zippy wrote the `HashMap` with unencrypted `Secret`s and I left it in and just renamed it to `cache` when adding the `BTreeMap`. I default to `BTreeMap` when it gets persisted after wasting hours in trying to figure out why DNA hashes were different, non-deterministically. It was the non-determinism of `HashMap`.

Reply:
:+1: